### PR TITLE
Issue 14583: [python client]: Improve garbage collection of producer/consumer/reader objects

### PR DESCRIPTION
**Pull request by zbentley**
Fixes https://github.com/apache/pulsar/issues/14583

### Motivation

The Python client bindings make it very easy to leak Pulsar client objects. Those leaks manifest in segfaults and program crashes (e.g. https://github.com/apache/pulsar/issues/14582). 

### Modifications

Adds an abstract superclass of all stateful python client objects (clients, consumers, readers, producers) which coordinates garbage collection of those objects, according to these principles:
- Disposing of an object should close its underlying resources.
- If the only reference a program holds is to a "child" object (one descended from a client) and that child is disposed of, the parent client that created it should also then be disconnected and destroyed.

### Verifying this change

This change added tests and can be verified as follows:
- Added a test case to `test_pulsar` which ensures that no file handles are leaked after consumer objects are destroyed. This test runs under an artificially lowered file handle limit; if handles are leaked, errors will be raised (well, segfaults and sigaborts are raised right now, but errors will be raised after #14582 is fixed).

### Documentation
- [X] `no-need-doc` 
I do not believe that this PR affects external API contracts or behavior.

### CODE REVIEW

**Review by merlimat — APPROVED** [review]

**Comment by BewareMyPower** [comment]
Could you fix the license check?

**Comment by zbentley** [comment]
Fixing license check and tests this week or next, sorry for delay.

File: pulsar-client-cpp/python/pulsar_test.py [hunk 1]
```diff
@@ -1228,6 +1228,16 @@ def test_json_schema_encode(self):
         second_encode = schema.encode(record)
         self.assertEqual(first_encode, second_encode)
 
+    def test_garbage_collection(self):
+        oldrlimit = resource.getrlimit(resource.RLIMIT_NOFILE)
+        try:
+            resource.setrlimit(resource.RLIMIT_NOFILE, (15, 15))
+            for i in range(100):
+                client = pulsar.Client(service_url=self.serviceUrl)
+                con = client.subscribe(topic='mytopic', subscription_name=f'mysub-{i}')
```
Comment (BewareMyPower) [inline pulsar-client-cpp/python/pulsar_test.py:1237]:
```suggestion
                con = client.subscribe(topic='mytopic', subscription_name='mysub-' + str(i))
```

Use a Python2 compatible way to create a string.

**Comment by BewareMyPower** [comment]
```
Traceback (most recent call last):
  File "pulsar_test.py", line 25, in <module>
    import pulsar
  File "/usr/local/lib/python2.7/dist-packages/pulsar/__init__.py", line 103, in <module>
    from abc import ABC
ImportError: cannot import name ABC
```

It looks like the `ABC` package doesn't exist in Python2, while the CI still uses Python2.

**Comment by merlimat** [comment]
> It looks like the `ABC` package doesn't exist in Python2, while the CI still uses Python2.

I think the py2 support at this time is out of any reasonable time :) 

I just submitted proposal to drop Py2. https://github.com/apache/pulsar/issues/15185. We should be able to do that in few days.

**Comment by zbentley** [comment]
Unsure why the two most recent commits haven't finished their build. Is there an issue with CI? Or a problem with this PR?

**Comment by BewareMyPower** [comment]
You can see error logs in  https://github.com/apache/pulsar/runs/6155320188?check_suite_focus=true.

```
Exception AttributeError: "'Client' object has no attribute '_handle'" in <bound method Client.__del__ of <pulsar.Client object at 0x7f0582d10ed0>> ignored
```

Debugging for the Python2 code is somehow troublesome. I usually build the Python2 wheels in a CentOS 7 docker with `docker-build.sh` and run tests inside the container.

If you are not in a hurry to merge this PR, maybe it's better to wait until PIP-155 is done. After that, you can restore this PR to the Python3 compatible code.

**Comment by zbentley** [comment]
@BewareMyPower there's no hurry. I'll fix up this code regardless since it needed a few tweaks. The build failures weren't due to that AttributeError, though (destruction time errors are ugly, but don't cause tests to fail with exceptions); they seemed to be due to the BROKER_CONNECTION, FUNCTION, and FLAKY builds.

**Comment by BewareMyPower** [comment]
@zbentley Could you resolve the conflicts in master branch?

P.S. This PR can only be cherry-picked into branch-2.11 because Pulsar drops the Python2 support since 2.11.0

**Comment by tisonkun** [comment]
Closed as the development of the Python client has been permanently moved to https://github.com/apache/pulsar-client-python.

**Pull request closed without merging**